### PR TITLE
Support safe_mode

### PR DESCRIPTION
**Pull request by beck**

### CODE REVIEW

**Comment by blaix** [comment]
do the tests pass with this change?

**Comment by beck** [comment]
Yep.  And technically this is reopening a pull @stefanor already thumbed (but I deleted the repo).  So I've already tagged these changes and they're already implemented in the relevant pull.

**Comment by beck** [comment]
Side note, stefano asked:

> Where's the description? Why is this forked? What's the plan for upstreaming these changes to get off the fork again?

I wrote a response but it is now lost (except for @stefanor's inbox)... maybe he will do me the favor of pasting it here for posterity

**Comment by stefanor** [comment]
@beck said:

> > Where's the description?
> 
> I was hoping the title, linking tickets, and comments was enough. Alas.
> As a followup to the the Definitions class pull, I told @bearnard I'd update the tests to also test in safe_mode. This is problematic because when the tests are run in isolation the deprecated warning causes tests to fail (sebastianbergmann#1710).
> This pull updates how phpunit opens a subprocess to ensure deprecated warnings are suppressed and that the subprocesses uses the same open_base dir and safe_mode settings.
> 
> > Why is this forked?
> 
> Because I thought this was the easiest way to review changes.
> 
> > What's the plan for upstreaming these changes to get off the fork again?
> 
> None.
> After reading through phpunit tickets, their sentiment is: "supporting legacy is a lot of work for few benefits....
> I'm not going to do it but if you can get it done nice and clean I don't mind having it". Which is a contradiction because supporting legacy is rarely "nice and clean".
> And he's right, trying to get safe_mode working has been grossly time consuming. I have no interest in pouring more time into trying to make this "nice and clean" (which I doubt is even possible).
> If a long-lived fork is not acceptable, another approach I considered:
> Since this <10 lines of code, store a patch next to the text script and after phpunit is composer installed, apply these changes

And then he had a follow-up, saying:

> This pull feels very specific to yola, instead of trying to get it pushed upstream, closing in favor of using a patch.

File: src/Util/PHP/Default.php [hunk 1]
```diff
@@ -34,9 +34,17 @@ class PHPUnit_Util_PHP_Default extends PHPUnit_Util_PHP
     public function runJob($job, array $settings = array())
     {
         $runtime = new Runtime;
+        $bin = $runtime->getBinary();
+
+        $settings[] = "error_reporting=" . (E_ALL ^ E_DEPRECATED);
+        if (ini_get('safe_mode')) {
+            $bin = str_replace("'", "", $bin);
+            $settings[] = "safe_mode=On";
+            $settings[] = "open_basedir=" . ini_get('open_basedir');
```
Comment (stefanor) [inline src/Util/PHP/Default.php]: This should probably be in a separate if block.

File: ChangeLog-Yola.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Changes made by Yola
+
+Always tag using `<sem version>-patch.#`.
```
Comment (stefanor) [inline ChangeLog-Yola.md]: I'd go with `-yola` rather than `-patch`. Makes it more obvious who is doing the patching.

Comment (stefanor) [inline ChangeLog-Yola.md]: Oh, that's what you meant by semantic versioning.

**Comment by stefanor** [comment]
:+1: again

**Comment by sebastianbergmann** [comment]
You do realize that `safe_mode` was removed in PHP 5.4?

**Comment by stefanor** [comment]
@sebastianbergmann: We produce PHP that runs on platforms outside our control, including some with `safe_mode` enabled. Until we can persuade those customers to get with the times, we have to ensure that this is a supported configuration for us, and unit tests are kind of handy for that.

**Pull request merged**